src/main/routing/network_graph.rs

```diff
@@ -0,0 +1,809 @@
+                assert_eq!(lookup_latency(node_2, node_1), 10);
+                assert_eq!(lookup_latency(node_2, node_2), 14);
+            }
+        }
+    }
+}
+
+mod export {
+    use super::*;
+
+    #[no_mangle]
+    pub extern "C" fn networkgraph_load(
+        config: *const configuration::ConfigOptions,
+    ) -> *mut NetworkGraph {
+        assert!(!config.is_null());
+        let config = unsafe { &*config };
+
+        match load_network_graph(config.network.graph.as_ref().unwrap()) {
+            Ok(graph_str) => Box::into_raw(Box::new(NetworkGraph::parse(&graph_str).unwrap())),
+            Err(err) => {
+                error!("{}", err);
+                std::ptr::null_mut()
+            }
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn networkgraph_free(graph: *mut NetworkGraph) {
+        assert!(!graph.is_null());
+        unsafe { Box::from_raw(graph) };
+    }
+
+    /// Get the downstream bandwidth of the graph node.
+    #[no_mangle]
+    pub extern "C" fn networkgraph_nodeBandwidthDownBits(
+        graph: *mut NetworkGraph,
+        node_id: u32,
+        bandwidth_down: *mut u64,
+    ) -> bool {
+        let graph = unsafe { graph.as_ref() }.unwrap();
+        let bandwidth_down = unsafe { bandwidth_down.as_mut() }.unwrap();
+
+        let node = graph.node_id_to_index(node_id).unwrap();
+
+        match graph.graph().node_weight(*node).unwrap().bandwidth_down {
+            Some(x) => {
+                *bandwidth_down = x.convert(units::SiPrefixUpper::Base).unwrap().value();
+                return true;
+            }
+            None => return false,
+        }
+    }
+
+    /// Get the upstream bandwidth of the graph node.
+    #[no_mangle]
+    pub extern "C" fn networkgraph_nodeBandwidthUpBits(
+        graph: *mut NetworkGraph,
+        node_id: u32,
+        bandwidth_up: *mut u64,
+    ) -> bool {
+        let graph = unsafe { graph.as_ref() }.unwrap();
+        let bandwidth_up = unsafe { bandwidth_up.as_mut() }.unwrap();
+
+        let node = graph.node_id_to_index(node_id).unwrap();
+
+        match graph.graph().node_weight(*node).unwrap().bandwidth_up {
+            Some(x) => {
+                *bandwidth_up = x.convert(units::SiPrefixUpper::Base).unwrap().value();
+                return true;
+            }
+            None => return false,
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn ipassignment_new() -> *mut IpAssignment<u32> {
+        Box::into_raw(Box::new(IpAssignment::new()))
+    }
+
+    #[no_mangle]
+    pub extern "C" fn ipassignment_free(ip_assignment: *mut IpAssignment<u32>) {
+        assert!(!ip_assignment.is_null());
+        unsafe { Box::from_raw(ip_assignment) };
+    }
+
+    /// Get an unused address and assign it to a node.
+    #[no_mangle]
+    pub extern "C" fn ipassignment_assignHost(
+        ip_assignment: *mut IpAssignment<u32>,
+        node_id: u32,
+        ip_addr: *mut libc::in_addr_t,
+    ) -> libc::c_int {
```

Comment:
Return bool for consistency with the other APIs? (Or change the others to return int?).

More generally I slightly prefer returning int and using either -1 or -errno as return values, for ~consistency with libc APIs. In particular though, mixing APIs with semantics of 'true means error' and 'non-zero means error' is especially error-prone in C since with implicit conversion to bool they have opposite semantics.

Reply:
Bool was only used for the bandwidth values to return if they were set or not, so I think anything that returns an error should stick with returning int. I think all of the other code follows the "return int and use -1 as error value" process.